components/eamxx/src/dynamics/homme/eamxx_homme_iop.cpp

```diff
@@ -0,0 +1,408 @@
+  uview_2d<Pack> delta_Q(delta_Q_slot.data(), n_q_tracers, nlev_packs);
+
+  auto s_pmid = ekat::scalarize(pmid);
+  auto s_omega = ekat::scalarize(omega);
+  auto s_delta_u = ekat::scalarize(delta_u);
+  auto s_delta_v = ekat::scalarize(delta_v);
+  auto s_delta_T = ekat::scalarize(delta_T);
+  auto s_delta_Q = ekat::scalarize(delta_Q);
+  auto s_omega_int = ekat::scalarize(omega_int);
+
+  // Compute omega on the interface grid by using a weighted average in pressure
+  const int pack_begin = 1/Pack::n, pack_end = (nlevs-1)/Pack::n;
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, pack_begin, pack_end+1), [&] (const int k){
+    auto range_pack = ekat::range<IntPack>(k*Pack::n);
+    range_pack.set(range_pack<1, 1);
+    Pack pmid_k, pmid_km1, omega_k, omega_km1;
+    ekat::index_and_shift<-1>(s_pmid, range_pack, pmid_k, pmid_km1);
+    ekat::index_and_shift<-1>(s_omega, range_pack, omega_k, omega_km1);
+
+    const auto weight = (pint(k) - pmid_km1)/(pmid_k - pmid_km1);
+    omega_int(k).set(range_pack>=1 and range_pack<=nlevs-1,
+                      weight*omega_k + (1-weight)*omega_km1);
+  });
+  omega_int(0)[0] = 0;
+  omega_int(nlevs/Pack::n)[nlevs%Pack::n] = 0;
+
+  // Compute delta views for u, v, T, and Q (e.g., u(k+1) - u(k), k=0,...,nlevs-2)
+  ColOps::compute_midpoint_delta(team, nlevs-1, u, delta_u);
+  ColOps::compute_midpoint_delta(team, nlevs-1, v, delta_v);
+  ColOps::compute_midpoint_delta(team, nlevs-1, T, delta_T);
+  for (int iq=0; iq<n_q_tracers; ++iq) {
+    auto tracer       = Kokkos::subview(Q,       iq, Kokkos::ALL());
+    auto delta_tracer = Kokkos::subview(delta_Q, iq, Kokkos::ALL());
+    ColOps::compute_midpoint_delta(team, nlevs-1, tracer, delta_tracer);
+  }
+  team.team_barrier();
+
+  // Compute updated temperature, horizontal winds, and tracers
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, nlev_packs), [&] (const int k) {
+    auto range_pack = ekat::range<IntPack>(k*Pack::n);
+
+    // Get delta(k-1) packs. We need a range pack
+    // that does not contain 0 so that we do not
+    // attempt to access k=-1 index.
+    auto range_pack_m1 = range_pack;
+    range_pack_m1.set(range_pack_m1<1, 1);
+    Pack delta_u_k, delta_u_km1,
+          delta_v_k, delta_v_km1,
+          delta_T_k, delta_T_km1;
+    ekat::index_and_shift<-1>(s_delta_u, range_pack_m1, delta_u_k, delta_u_km1);
+    ekat::index_and_shift<-1>(s_delta_v, range_pack_m1, delta_v_k, delta_v_km1);
+    ekat::index_and_shift<-1>(s_delta_T, range_pack_m1, delta_T_k, delta_T_km1);
+
+    // Get omega_int(k+1) pack. We don't need a specialized
+    // range pack since omega_int contains nlevs+1 entries.
+    Pack omega_int_k, omega_int_kp1;
+    ekat::index_and_shift<1>(s_omega_int, range_pack, omega_int_k, omega_int_kp1);
+
+    auto at_top = range_pack==0;
+    auto at_bot = range_pack==nlevs-1;
+
+    auto fac = dt/(2*pdel(k));
+
+    // Update u
+    u(k).set(at_top,                    u(k) - fac*omega_int_kp1*delta_u_k);
```

Comment:
One general thing for this for loop. I would consider grabbing a ref out of the view, so you don't index the view 3+ times. Something like
```suggestion
auto& uk = u(k);
uk.set(...);
uk.set(...);
uk.set(...);
```